Sources/SwiftLanguageService/SignatureHelp.swift

```diff
@@ -0,0 +1,145 @@
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+package import LanguageServerProtocol
+import SKLogging
+import SourceKitD
+import SourceKitLSP
+import SwiftBasicFormat
+
+fileprivate extension String {
+  func utf16Offset(of utf8Offset: Int, callerFile: StaticString = #fileID, callerLine: UInt = #line) -> Int {
+    guard
+      let stringIndex = self.utf8.index(self.startIndex, offsetBy: utf8Offset, limitedBy: self.endIndex)
+    else {
+      logger.fault(
+        """
+        UTF-8 offset is past the end of the string while getting UTF-16 offset of \(utf8Offset) \
+        (\(callerFile, privacy: .public):\(callerLine, privacy: .public))
+        """
+      )
+      return self.utf16.count
+    }
+    return self.utf16.distance(from: self.startIndex, to: stringIndex)
+  }
+}
+
+fileprivate extension ParameterInformation {
+  init?(_ parameter: SKDResponseDictionary, _ signatureLabel: String, _ keys: sourcekitd_api_keys) {
+    guard let nameOffset = parameter[keys.nameOffset] as Int?,
+      let nameLength = parameter[keys.nameLength] as Int?
+    else {
+      return nil
+    }
+
+    let documentation: StringOrMarkupContent? =
+      if let docComment: String = parameter[keys.docComment] {
+        .markupContent(MarkupContent(kind: .markdown, value: docComment))
+      } else {
+        nil
+      }
+
+    let labelStart = signatureLabel.utf16Offset(of: nameOffset)
+    let labelEnd = signatureLabel.utf16Offset(of: nameOffset + nameLength)
+
+    self.init(
+      label: .offsets(start: labelStart, end: labelEnd),
+      documentation: documentation
+    )
+  }
+}
+
+fileprivate extension SignatureInformation {
+  init?(_ signature: SKDResponseDictionary, _ keys: sourcekitd_api_keys) {
+    guard let label = signature[keys.name] as String?,
+      let skParameters = signature[keys.parameters] as SKDResponseArray?
+    else {
+      return nil
+    }
+
+    let parameters = skParameters.compactMap { ParameterInformation($0, label, keys) }
+
+    let activeParameter: Int? =
+      if let activeParam: Int = signature[keys.activeParameter] {
+        activeParam
+      } else if !parameters.isEmpty {
+        // If we have parameters and no active parameter is present, we return
+        // an out-of-range index that way editors don't show an active parameter.
+        // As of LSP 3.17, not returning an active parameter defaults to choosing
+        // the first parameter which isn't the desired behavior.
+        // LSP 3.18 defines a `noActiveParameterSupport` option which allows the
+        // active parameter to be `null` causing editors not to show an active
+        // parameter which would be the best solution here.
+        parameters.count
```

Comment:
Oh, that behavior is interesting. The LSP specification says the following. Does this mean that VS Code does not comply with the LSP spec in this regard?

> If omitted or the value lies outside the range of `signatures[activeSignature].parameters` defaults to 0 if the active signature has parameters

I’m fine keeping this in but we should mention the intended behavior from the LSP spec in the comment and mention that we are deliberately relying on a VS Code LSP-non-compliance if that’s what we’re doing.

Reply:
Yes, that's VS Code-specific behavior which we use as a workaround here. For editors that are fully LSP-compliant this will have the same behavior as if we didn't return a value and we actually don't have a way of expressing "no active parameter" in that case so it's fine. I actually found rust-analyzer using this trick too.

There are also other VS Code-specific workarounds like setting the active parameter to `-1` (https://github.com/microsoft/vscode/issues/145851#issuecomment-1078689732).

I'll update the comment to explain this part too.